### PR TITLE
Button editor

### PR DESCRIPTION
**Pull request by dkrasner**
### Main points ###

After some pain, a button editor. It does the basic things: 

Changes the name
Adjusts background and font colors (via our colorWheel)
Lists, adds, removes events
And lets you script

image

The editor is opened by pressing the "write/edit" icon, which previously opened the script/field editor. 

Instead of embedding all the widgets necessary to do editing (color, script) inside one editor view I decided that each would open it's own, absolutely positioned, widget. After some experimenting with both options, this was easier to style and scale, and in the end I prefer it overall. This way you can lay out the various sub-editor widgets as you see fit and we can just keep making more, adding  a button to open it to the main one:

image

I tried to [stick to SimpleTalk](https://github.com/UnitedLexCorp/SimpleTalk/compare/daniel-button-editor?expand=1#diff-d460e66340863153b40b711a201ef3dea6540293e5d5bc796755e4f238f0a473R241) in evoking the various sub-editors so that it would be clearer how one could build such a thing with just SimpleTalk buttons. You almost can do it at the moment... 

`editorOpen` is a now a [property](https://github.com/UnitedLexCorp/SimpleTalk/compare/daniel-button-editor?expand=1#diff-d460e66340863153b40b711a201ef3dea6540293e5d5bc796755e4f238f0a473R241) on all parts. 

#### Tests ####

I wrote as many [tests](https://github.com/UnitedLexCorp/SimpleTalk/compare/daniel-button-editor?expand=1#diff-0a1b715b2d04f9c839158dcb7132bfb2ba73b511ea8d03a2097b1851ac0dbe49) as I could think of, without dealing with npm canvas context issues (which you need to solve if you want to properly test our color wheels). They are reasonably comprehensive. 

#### Some issues ####
The fact that we couldn't write this in ST is a bit lame.. but I think we can get there. 
The css here is a mess. 
Adding and removing events causes this issue #97  (nothing to do with editor...)
The editor view is not really responsive. For example, it can send a `respond to [event]` message but it is not aware that the property changed to respond to that, so you have to [manually add it](https://github.com/UnitedLexCorp/SimpleTalk/compare/daniel-button-editor?expand=1#diff-d460e66340863153b40b711a201ef3dea6540293e5d5bc796755e4f238f0a473R320). This is easy enough here, but we know in more complicated situations the fact that there is no responsiveness between the target model and it's editor. Potentially having the editor add itself to the target and then using the `target.onPropChange()` could work. But then we'd have `editor.target` and `target.editor` attributes floating around... or add a model for the editor. This is the reason I left a [basic version](https://github.com/UnitedLexCorp/SimpleTalk/compare/daniel-button-editor?expand=1#diff-5c04e967413aec14e8cd22a4fe1f520620357e5acb7c9e24d4c5eb879fc22674) of that in (at first I had an editor model/view setup). TBD... 

Give it a go...

### CODE REVIEW

File: js/objects/views/editors/ButtonEditorView.js [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+        // TODO: this is a hack b/c the editor is not properly responsive at the moment
+        let eventEl = eventListDiv.querySelector(`#${eventName}`);
+        if(eventEl){
+            return;
+        }
+        eventEl = document.createElement("div");
+        let eventSpan = document.createElement("span");
+        let closeSpan = document.createElement("span");
+        let xmlDocument = SVGParser.parseFromString(closeIcon, 'application/xml');
+        let closeSvgEl = xmlDocument.documentElement;
```
Comment (darth-cheney) [inline js/objects/views/editors/ButtonEditorView.js:363]: Where does this get used/displayed?

**Review by darth-cheney — COMMENTED** [review]
This looks good to me. Just a couple of points/issues:
1. Changing the button's text color also changes the color of the icons in the open halo, as in this screenshot: 
Screenshot from 2021-01-28 12-23-40
  
2. I can no longer adjust the size of the script editor window using the bottom right corner. This seems to trigger strange dragging behavior
  
Once we get some more of the critical parts of the system down, we should come back and explore the buttons in this editor. I was confused at first about how to remove an event listener for a button (the - sign is very small and the only clickable part of it). We might also consider an autocompletion dropdown for adding the event listeners which says the events available for the given Part.

File: js/objects/views/editors/ButtonEditorView.js (hunk 1 above)
Comment (dkrasner) [reply js/objects/views/editors/ButtonEditorView.js:363]: ah that's crud - will remove

**Comment by dkrasner** [comment]
> This looks good to me. Just a couple of points/issues:
> 
>     1. Changing the button's text color also changes the color of the icons in the open halo, as in this screenshot:
>        Screenshot from 2021-01-28 12-23-40
> 

yeah i noticed that, but then like it but can remove; it's just a css thing on the halo which need to be explicit i think..


>     2. I can no longer adjust the size of the script editor window using the bottom right corner. This seems to trigger strange dragging behavior
ok i'll look into it, but i think this is already a problem in `master`  in which case i'll fix it in another PR

> 
> Once we get some more of the critical parts of the system down, we should come back and explore the buttons in this editor. I was confused at first about how to remove an event listener for a button (the - sign is very small and the only clickable part of it). We might also consider an autocompletion dropdown for adding the event listeners which says the events available for the given Part.
yeah it's shitty, i wasn't sure how we want to lay these out or whether we will the fact that the editor is not one thing but more an interface to open various sub-editors. If we like the latter, maybe we can just open an `events` sub-editor and have that nicely styled etc

**Pull request merged**